webui/react/src/components/UPlot/UPlotChart.tsx

```diff
@@ -26,141 +25,208 @@ interface ScaleZoomData {
   min?: number;
 }
 
+const shouldUpdate = (
+  prev: Partial<uPlot.Options> | undefined,
+  next: Partial<uPlot.Options> | undefined,
+  chart: uPlot | undefined,
+): boolean => {
+  if (!chart || !next) return false;
+  if (!prev) return true;
+  if (Object.keys(prev).length !== Object.keys(next).length) {
+    return true;
+  }
+  if (prev.axes?.length !== next.axes?.length) {
+    return true;
+  }
+
+  const someAxisLabelHasChanged =
+    prev.axes?.some((prevAxis, seriesIdx) => {
+      const nextAxis = next.axes?.[seriesIdx];
+      return prevAxis.label !== nextAxis?.label;
+    });
+  if (someAxisLabelHasChanged) return true;
+
+  if (chart?.series?.length !== next.series?.length) {
+    return true;
+  }
+
+  if (prev.scales?.y?.distr !== next.scales?.y?.distr) return true;
+
+  const someSeriesHasChanged =
+    chart.series.some((chartSerie, seriesIdx) => {
+      const nextSerie = next.series?.[seriesIdx];
+      const prevSerie = prev.series?.[seriesIdx];
+      return (
+        (nextSerie?.show != null && chartSerie?.show !== nextSerie?.show)
+        || (prevSerie?.show != null && prevSerie?.show !== nextSerie?.show)
+        || (nextSerie?.label != null && chartSerie?.label !== nextSerie?.label)
+        || (nextSerie?.fill != null && chartSerie?.fill !== nextSerie?.fill)
+      );
+    });
+  if(someSeriesHasChanged) return true;
+
+  return false;
+};
+
+const getNormalizedData = (
```

Comment:
**nit suggestion:** my thinking in [my PR](https://github.com/determined-ai/determined/pull/3976/files#diff-b913991799a10b7076974f349ae6d960b8982b5d97f829cdf2e9fcd64a2c4f8fR131-R151):
- move it close to where it's used since it's not used anywhere else
- the effect is responsible for taking data prop updates and do some processing and updating the dataRef, seemed small of a code enough to keep it close to source